analysis/03_new_model_training/03_model_training.md

```diff
@@ -0,0 +1,561 @@
+    mse_train_idx3 = mean_squared_error(
+        y_train[bin_index_train == 3], y_pred_train[bin_index_train == 3]
+    )
+    rmse_train_3 = np.sqrt(mse_train_idx3)
+    mse_train_idx4 = mean_squared_error(
+        y_train[bin_index_train == 4], y_pred_train[bin_index_train == 4]
+    )
+    rmse_train_4 = np.sqrt(mse_train_idx4)
+    mse_train_idx5 = mean_squared_error(
+        y_train[bin_index_train == 5], y_pred_train[bin_index_train == 5]
+    )
+    rmse_train_5 = np.sqrt(mse_train_idx5)
+
+    # Estimation of RMSE for test data
+    y_pred = xgb.predict(X_test)
+
+    mse_idx1 = mean_squared_error(
+        y_test[bin_index_test == 1], y_pred[bin_index_test == 1]
+    )
+    rmse_1 = np.sqrt(mse_idx1)
+    mse_idx2 = mean_squared_error(
+        y_test[bin_index_test == 2], y_pred[bin_index_test == 2]
+    )
+    rmse_2 = np.sqrt(mse_idx2)
+    mse_idx3 = mean_squared_error(
+        y_test[bin_index_test == 3], y_pred[bin_index_test == 3]
+    )
+    rmse_3 = np.sqrt(mse_idx3)
+    mse_idx4 = mean_squared_error(
+        y_test[bin_index_test == 4], y_pred[bin_index_test == 4]
+    )
+    rmse_4 = np.sqrt(mse_idx4)
+    mse_idx5 = mean_squared_error(
+        y_test[bin_index_test == 5], y_pred[bin_index_test == 5]
+    )
+    rmse_5 = np.sqrt(mse_idx5)
+
+    train_RMSE_list_bin1.append(rmse_train_1)
+    train_RMSE_list_bin2.append(rmse_train_2)
+    train_RMSE_list_bin3.append(rmse_train_3)
+    train_RMSE_list_bin4.append(rmse_train_4)
+    train_RMSE_list_bin5.append(rmse_train_5)
+
+    test_RMSE_list_bin1.append(rmse_1)
+    test_RMSE_list_bin2.append(rmse_2)
+    test_RMSE_list_bin3.append(rmse_3)
+    test_RMSE_list_bin4.append(rmse_4)
+    test_RMSE_list_bin5.append(rmse_5)
+```
+
+## Plot RMSE in total
+
+```python
+# RMSE in total
+
+m_test_rmse = statistics.mean(test_RMSE_list)
+plt.figure(figsize=(4, 3))
+plt.axvline(m_test_rmse, color="red", linestyle="dashed")
+plt.hist(
+    test_RMSE_list,
+    bins=np.arange(12.0, 14.0, 0.12),
+    edgecolor="k",
+    histtype="bar",
+    density=True,
+)
+sd_test_rmse = statistics.stdev(test_RMSE_list)
+
+
+m_train_rmse = statistics.mean(train_RMSE_list)
+plt.axvline(m_train_rmse, color="b", linestyle="dashed")
+plt.hist(
+    train_RMSE_list,
+    bins=np.arange(12.0, 14.0, 0.12),
+    color="orange",
+    edgecolor="k",
+    histtype="bar",
+    density=True,
+    alpha=0.7,
+)
+sd_train_rmse = statistics.stdev(train_RMSE_list)
+
+print(f"stdev_RMSE_test: {sd_test_rmse:.2f}")
+print(f"stdev_RMSE_train: {sd_train_rmse:.2f}")
+
+print(f"mean_RMSE_test: {m_test_rmse:.2f}")
+print(f"mean_RMSE_train: {m_train_rmse:.2f}")
+
+# create legend
+labels = ["Mean_test", "Mean_train", "test", "train"]
+plt.legend(labels)
+
+plt.xlabel("The RMSE error")
+plt.ylabel("Frequency")
+plt.title("histogram of the RMSE distribution")
+plt.show()
+```
+
+## Plot RMSE per bins
+
+```python
+# RMSE of bins_1
+
+m_test_rmse = statistics.mean(test_RMSE_list_bin1)
```

Comment:
All this code is repeated for each bin, just make a function and call it 5 times